File: Python/332. Reconstruct Itinerary.py
'''
回溯：横向遍历目的地，纵横向递归出发地
'''


class Solution:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        tickets_dict = collections.defaultdict(list)
        for item in tickets:
            tickets_dict[item[0]].append(item[1])
        path = ["JFK"]

        def backtrack(start, numTickets):
            nonlocal tickets_dict
            nonlocal path
            if len(path) == numTickets + 1:
                return True
            tickets_dict[start].sort()
            for i in range(len(tickets_dict[start])):
                des = tickets_dict[start].pop(0)
                path.append(des)
                if backtrack(des, numTickets):
                    return True
                des = path.pop()
                tickets_dict[start].append(des)
            return False

        backtrack("JFK", len(tickets))
        return path

File: C++/332. Reconstruct Itinerary.cpp
/*
回溯：横向遍历目的地，纵横向递归出发地
*/
class Solution {
public:
    unordered_map<string, map<string, int>> targets;
    vector<string> path;
    bool backtrack(const string& from, int ticketNum) {
        if (path.size() == ticketNum + 1) {
            return true;
        }
        for (pair<const string, int>& target : targets[from]) {
            if (target.second > 0) {
                path.push_back(target.first);
                target.second--;
                if (backtrack(target.first, ticketNum)) return true;
                target.second++;
                path.pop_back();
            }
        }
        return false;
    }
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        targets.clear();
        path.clear();
        for (auto ticket : tickets) {
            targets[ticket[0]][ticket[1]]++;
        }
        path.push_back("JFK");
        backtrack("JFK", tickets.size());
        return path;
    }
};